File: results/2026-04-08_192624/tasks/11-semantic-version-bumper/default-sonnet/generated-code/run_tests.py
#!/usr/bin/env python3
"""
Test harness for the semantic version bumper.

For each test case this script:
  1. Creates an isolated temp git repository.
  2. Copies project files (bumper.py, test_bumper.py, workflow) into it.
  3. Seeds a package.json with the case's initial version and a fixtures/commits.txt
     with mock commit messages.
  4. Runs `act push --rm` inside that repo and captures all output.
  5. Appends the output (clearly delimited) to act-result.txt in the original CWD.
  6. Asserts act exited with code 0 and the output contains the expected new version.
  7. Asserts every job shows "Job succeeded".

All failures are collected; the script exits non-zero if any case fails.
"""

import json
import os
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import List, NamedTuple, Optional

# ─── Project root (where this script lives) ───────────────────────────────────

PROJECT_DIR = Path(__file__).parent.resolve()

# Output file written to the original CWD
ACT_RESULT_FILE = PROJECT_DIR / "act-result.txt"


# ─── Test-case definitions ─────────────────────────────────────────────────────

class TestCase(NamedTuple):
    name: str
    description: str
    initial_version: str
    commits: List[str]          # one mock commit message per entry
    expected_version: str       # exact string that must appear in act output


TEST_CASES: List[TestCase] = [
    TestCase(
        name="patch-bump",
        description="fix commits only → patch bump",
        initial_version="1.0.0",
        commits=[
            "fix: resolve null pointer exception",
            "fix(ui): correct button alignment",
            "chore: update dependencies",
        ],
        expected_version="1.0.1",
    ),
    TestCase(
        name="minor-bump",
        description="feat commit → minor bump (overrides fix)",
        initial_version="1.0.0",
        commits=[
            "feat: add new dashboard",
            "fix: minor bug fix",
        ],
        expected_version="1.1.0",
    ),
    TestCase(
        name="major-bump",
        description="breaking-change commit → major bump",
        initial_version="1.0.0",
        commits=[
            "feat!: redesign API interface",
            "fix: minor correction",
        ],
        expected_version="2.0.0",
    ),
    TestCase(
        name="no-bump",
        description="non-conventional commits → version unchanged",
        initial_version="1.2.3",
        commits=[
            "chore: update dependencies",
            "docs: update readme",
            "ci: fix workflow",
        ],
        expected_version="1.2.3",
    ),
    TestCase(
        name="minor-overrides-patch",
        description="feat + fix together → minor wins",
        initial_version="1.1.0",
        commits=[
            "feat: add user profile page",
            "fix: fix login timeout",
        ],
        expected_version="1.2.0",
    ),
]


# ─── Helpers ──────────────────────────────────────────────────────────────────

def run(cmd: List[str], cwd: Optional[str] = None, check: bool = True) -> subprocess.CompletedProcess:
    """Run a command, optionally raising on failure."""
    return subprocess.run(cmd, cwd=cwd, capture_output=True, text=True, check=check)


def setup_temp_repo(case: TestCase) -> Path:
    """
    Create a temporary git repository pre-loaded with project files and
    fixture data for the given test case.

    Structure:
        <tmp>/
          bumper.py              ← copied from PROJECT_DIR
          test_bumper.py         ← copied from PROJECT_DIR
          package.json           ← seeds initial_version
          fixtures/
            commits.txt          ← mock commit messages (one per line)
          .github/workflows/
            semantic-version-bumper.yml  ← copied from PROJECT_DIR
    """
    tmp = Path(tempfile.mkdtemp(prefix=f"act-test-{case.name}-"))

    # ── git init ──────────────────────────────────────────────────────────────
    run(["git", "init", "-b", "main"], cwd=str(tmp))
    run(["git", "config", "user.email", "ci@test.local"], cwd=str(tmp))
    run(["git", "config", "user.name", "CI Test"], cwd=str(tmp))

    # ── Copy project files ────────────────────────────────────────────────────
    shutil.copy2(PROJECT_DIR / "bumper.py",      tmp / "bumper.py")
    shutil.copy2(PROJECT_DIR / "test_bumper.py", tmp / "test_bumper.py")

    workflow_src = PROJECT_DIR / ".github" / "workflows" / "semantic-version-bumper.yml"
    workflow_dst = tmp / ".github" / "workflows"
    workflow_dst.mkdir(parents=True)
    shutil.copy2(workflow_src, workflow_dst / "semantic-version-bumper.yml")

    # ── Seed package.json ─────────────────────────────────────────────────────
    pkg = {"name": "test-project", "version": case.initial_version, "private": True}
    (tmp / "package.json").write_text(json.dumps(pkg, indent=2) + "\n")

    # ── Fixture: commits.txt (one commit message per line) ────────────────────
    fixtures_dir = tmp / "fixtures"
    fixtures_dir.mkdir()
    (fixtures_dir / "commits.txt").write_text("\n".join(case.commits) + "\n")

    # ── Initial commit so git is happy ────────────────────────────────────────
    run(["git", "add", "."], cwd=str(tmp))
    run(["git", "commit", "-m", "chore: initial project setup"], cwd=str(tmp))

    return tmp


def run_act(repo_dir: Path) -> subprocess.CompletedProcess:
    """
    Execute `act push --rm` inside repo_dir and return the completed process.
    stdout+stderr are captured but NOT checked here — caller decides pass/fail.
    """
    return subprocess.run(
        [
            "act", "push",
            "--rm",
            "--no-cache-server",
            # Use the pre-pulled image to avoid network fetches
            "-P", "ubuntu-latest=catthehacker/ubuntu:act-latest",
        ],
        cwd=str(repo_dir),
        capture_output=True,
        text=True,
        timeout=300,   # 5 min safety net
    )


def check_assertions(case: TestCase, result: subprocess.CompletedProcess) -> List[str]:
    """
    Return a list of failure messages.  Empty list means all assertions passed.
    """
    failures: List[str] = []
    combined = result.stdout + result.stderr

    # 1. Exit code 0
    if result.returncode != 0:
        failures.append(f"act exited with code {result.returncode} (expected 0)")

    # 2. Expected new version present in output
    expected_tag = f"NEW_VERSION={case.expected_version}"
    if expected_tag not in combined:
        failures.append(
            f"Expected '{expected_tag}' in output, not found.\n"
            f"  (searched stdout+stderr for exact match)"
        )

    # 3. "Job succeeded" present
    if "Job succeeded" not in combined:
        failures.append("Expected 'Job succeeded' in act output, not found")

    return failures


# ─── Main ─────────────────────────────────────────────────────────────────────

def main() -> int:
    print(f"Running {len(TEST_CASES)} test case(s) through act…\n")

    # Truncate the result file for this run
    ACT_RESULT_FILE.write_text(
        f"=== act-result.txt — Semantic Version Bumper Test Run ===\n\n"
    )

    all_passed = True

    for idx, case in enumerate(TEST_CASES, start=1):
        print(f"[{idx}/{len(TEST_CASES)}] {case.name}: {case.description}")

        repo_dir = setup_temp_repo(case)

        try:
            print(f"  → Running act in {repo_dir} …", flush=True)
            result = run_act(repo_dir)
        except subprocess.TimeoutExpired:
            result = subprocess.CompletedProcess(
                args=[], returncode=1,
                stdout="", stderr="act timed out after 300 s",
            )

        # ── Append to act-result.txt ──────────────────────────────────────────
        delimiter = "=" * 72
        with ACT_RESULT_FILE.open("a") as f:
            f.write(f"{delimiter}\n")
            f.write(f"TEST CASE {idx}: {case.name}\n")
            f.write(f"  description : {case.description}\n")
            f.write(f"  initial ver : {case.initial_version}\n")
            f.write(f"  expected ver: {case.expected_version}\n")
            f.write(f"  commits     : {case.commits}\n")
            f.write(f"{delimiter}\n")
            f.write("--- STDOUT ---\n")
            f.write(result.stdout or "(empty)\n")
            f.write("--- STDERR ---\n")
            f.write(result.stderr or "(empty)\n")
            f.write(f"--- EXIT CODE: {result.returncode} ---\n\n")

        # ── Assertions ────────────────────────────────────────────────────────
        failures = check_assertions(case, result)

        if failures:
            all_passed = False
            print(f"  ✗ FAILED:")
            for msg in failures:
                print(f"    • {msg}")
        else:
            print(f"  ✓ PASSED (NEW_VERSION={case.expected_version})")

        # ── Clean up temp repo ────────────────────────────────────────────────
        shutil.rmtree(repo_dir, ignore_errors=True)

        print()

    # ── Final summary ─────────────────────────────────────────────────────────
    with ACT_RESULT_FILE.open("a") as f:
        f.write("=" * 72 + "\n")
        status = "ALL PASSED" if all_passed else "SOME FAILED"
        f.write(f"OVERALL RESULT: {status}\n")
        f.write("=" * 72 + "\n")

    if all_passed:
        print("✓ All test cases passed.")
        print(f"Results written to: {ACT_RESULT_FILE}")
        return 0
    else:
        print("✗ One or more test cases FAILED.")
        print(f"Results written to: {ACT_RESULT_FILE}")
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: results/2026-05-06_173435/tasks/11-semantic-version-bumper/typescript-bun-opus/generated-code/run-tests.ts
#!/usr/bin/env bun
// Test harness that runs all test cases through act (GitHub Actions).
// Creates temp repos for each fixture, runs act, captures output,
// validates exact expected values.

import { mkdtempSync, writeFileSync, readFileSync, cpSync, rmSync, existsSync, mkdirSync } from "fs";
import { join } from "path";
import { tmpdir } from "os";

interface TestCase {
  name: string;
  fixture: string;
  initialVersion: string;
  expectedVersion: string;
  expectedBump: string;
  expectBumped: boolean;
}

const PROJECT_DIR = import.meta.dir;

const testCases: TestCase[] = [
  {
    name: "feat-commits-minor-bump",
    fixture: "fixtures/commits-feat.txt",
    initialVersion: "1.1.0",
    expectedVersion: "1.2.0",
    expectedBump: "minor",
    expectBumped: true,
  },
  {
    name: "fix-commits-patch-bump",
    fixture: "fixtures/commits-fix.txt",
    initialVersion: "2.3.1",
    expectedVersion: "2.3.2",
    expectedBump: "patch",
    expectBumped: true,
  },
  {
    name: "breaking-commits-major-bump",
    fixture: "fixtures/commits-breaking.txt",
    initialVersion: "1.5.3",
    expectedVersion: "2.0.0",
    expectedBump: "major",
    expectBumped: true,
  },
];

function setupTempRepo(tc: TestCase): string {
  const tempDir = mkdtempSync(join(tmpdir(), `svb-test-${tc.name}-`));

  // Copy source files
  const filesToCopy = [
    "version-bumper.ts",
    "bump-version.ts",
    "package.json",
    "tsconfig.json",
  ];
  for (const f of filesToCopy) {
    cpSync(join(PROJECT_DIR, f), join(tempDir, f));
  }

  // Copy workflow
  mkdirSync(join(tempDir, ".github", "workflows"), { recursive: true });
  cpSync(
    join(PROJECT_DIR, ".github", "workflows", "semantic-version-bumper.yml"),
    join(tempDir, ".github", "workflows", "semantic-version-bumper.yml")
  );

  // Copy fixtures dir
  mkdirSync(join(tempDir, "fixtures"), { recursive: true });
  cpSync(join(PROJECT_DIR, tc.fixture), join(tempDir, "commit-log.txt"));

  // Copy all fixtures for the test run
  for (const f of ["commits-feat.txt", "commits-fix.txt", "commits-breaking.txt", "commits-none.txt"]) {
    cpSync(join(PROJECT_DIR, "fixtures", f), join(tempDir, "fixtures", f));
  }

  // Copy test file
  cpSync(join(PROJECT_DIR, "version-bumper.test.ts"), join(tempDir, "version-bumper.test.ts"));

  // Write VERSION file with initial version
  writeFileSync(join(tempDir, "VERSION"), tc.initialVersion + "\n");

  // Write .actrc
  writeFileSync(join(tempDir, ".actrc"), "-P ubuntu-latest=act-ubuntu-pwsh:latest\n");

  // Copy bun lockfile if exists
  const lockPath = join(PROJECT_DIR, "bun.lock");
  if (existsSync(lockPath)) {
    cpSync(lockPath, join(tempDir, "bun.lock"));
  }

  // Initialize a git repo (needed for actions/checkout)
  Bun.spawnSync(["git", "init"], { cwd: tempDir });
  Bun.spawnSync(["git", "config", "user.email", "test@test.com"], { cwd: tempDir });
  Bun.spawnSync(["git", "config", "user.name", "Test"], { cwd: tempDir });
  Bun.spawnSync(["git", "add", "."], { cwd: tempDir });
  Bun.spawnSync(["git", "commit", "-m", "initial"], { cwd: tempDir });

  return tempDir;
}

function runAct(tempDir: string): { exitCode: number; output: string } {
  const result = Bun.spawnSync(["act", "push", "--rm", "--pull=false"], {
    cwd: tempDir,
    timeout: 300_000,
    env: { ...process.env, HOME: process.env.HOME || "/root" },
  });
  const stdout = result.stdout.toString();
  const stderr = result.stderr.toString();
  return {
    exitCode: result.exitCode,
    output: stdout + "\n" + stderr,
  };
}

function main(): void {
  const resultFile = join(PROJECT_DIR, "act-result.txt");
  let allOutput = "";
  let allPassed = true;
  const failures: string[] = [];

  console.log(`Running ${testCases.length} test cases through act...\n`);

  for (const tc of testCases) {
    console.log(`--- Test: ${tc.name} ---`);
    const tempDir = setupTempRepo(tc);
    console.log(`  Temp dir: ${tempDir}`);

    const { exitCode, output } = runAct(tempDir);

    allOutput += `\n${"=".repeat(60)}\n`;
    allOutput += `TEST CASE: ${tc.name}\n`;
    allOutput += `Initial version: ${tc.initialVersion}\n`;
    allOutput += `Expected version: ${tc.expectedVersion}\n`;
    allOutput += `${"=".repeat(60)}\n`;
    allOutput += output;

    // Assert exit code
    if (exitCode !== 0) {
      const msg = `FAIL [${tc.name}]: act exited with code ${exitCode}`;
      console.error(`  ${msg}`);
      failures.push(msg);
      allPassed = false;
      // Clean up and continue
      rmSync(tempDir, { recursive: true, force: true });
      continue;
    }
    console.log(`  act exit code: 0 (OK)`);

    // Assert job succeeded
    if (!output.includes("Job succeeded")) {
      const msg = `FAIL [${tc.name}]: "Job succeeded" not found in output`;
      console.error(`  ${msg}`);
      failures.push(msg);
      allPassed = false;
    } else {
      console.log(`  Job succeeded: YES`);
    }

    // Assert exact expected version in output
    if (tc.expectBumped) {
      if (!output.includes(`New version: ${tc.expectedVersion}`)) {
        const msg = `FAIL [${tc.name}]: Expected "New version: ${tc.expectedVersion}" not found`;
        console.error(`  ${msg}`);
        failures.push(msg);
        allPassed = false;
      } else {
        console.log(`  Version output matches: ${tc.expectedVersion}`);
      }

      // Assert bump type
      if (!output.includes(`Bump type: ${tc.expectedBump}`)) {
        const msg = `FAIL [${tc.name}]: Expected "Bump type: ${tc.expectedBump}" not found`;
        console.error(`  ${msg}`);
        failures.push(msg);
        allPassed = false;
      } else {
        console.log(`  Bump type matches: ${tc.expectedBump}`);
      }

      // Assert VERSION file content shown in results
      if (!output.includes(tc.expectedVersion)) {
        const msg = `FAIL [${tc.name}]: Expected version "${tc.expectedVersion}" not in output`;
        console.error(`  ${msg}`);
        failures.push(msg);
        allPassed = false;
      }
    }

    // Assert unit tests passed
    if (!output.includes("29 pass") && !output.includes("pass")) {
      const msg = `FAIL [${tc.name}]: Unit tests did not pass`;
      console.error(`  ${msg}`);
      failures.push(msg);
      allPassed = false;
    } else {
      console.log(`  Unit tests: PASSED`);
    }

    console.log(`  RESULT: ${failures.length === 0 ? "PASS" : "FAIL"}\n`);

    // Clean up temp dir
    rmSync(tempDir, { recursive: true, force: true });
  }

  // Write all output to act-result.txt
  writeFileSync(resultFile, allOutput);
  console.log(`\nResults written to: ${resultFile}`);

  if (!allPassed) {
    console.error(`\n${failures.length} assertion(s) failed:`);
    for (const f of failures) {
      console.error(`  - ${f}`);
    }
    process.exit(1);
  }

  console.log(`\nAll ${testCases.length} test cases PASSED!`);
}

main();
